fabtests/common/shared.c

```diff
@@ -2154,21 +2154,25 @@ ft_tag_is_valid(struct fid_cq * cq, struct fi_cq_err_entry *comp, uint64_t tag)
  * fi_cq_err_entry can be cast to any CQ entry format.
  */
 static int ft_spin_for_comp(struct fid_cq *cq, uint64_t *cur,
-			    uint64_t total, int timeout)
+			    uint64_t total, int timeout,
+			    struct fi_cq_err_entry *cq_entries)
 {
-	struct fi_cq_err_entry comp;
+	struct fi_cq_err_entry comp, *comp_ptr;
 	struct timespec a, b;
+	int bgn;
 	int ret;
 
 	if (timeout >= 0)
 		clock_gettime(CLOCK_MONOTONIC, &a);
 
+	bgn = *cur;
 	do {
-		ret = fi_cq_read(cq, &comp, 1);
+		comp_ptr = cq_entries ? &cq_entries[*cur - bgn] : &comp;
```

Comment:
If we're going to make this change, don't make the parameter optional.  Require callers to pass in values.

But these changes look too complex.  They are requiring the caller to allocate an array of a variable size (total - cur) prior to calling the function.  This sort of change needs to be separated out, so it can be examined more closely.

My guess is that the caller only needs to get a returned list of contexts that were freed and can be re-used.  It may be possible to free those entries here instead.

That the same set of changes is being made in 3 places is a sure sign that the current code should be simplified prior to layering on these changes.  So that only a single update is needed.  For example, introduce a 'handle_comp' function that can be called after fi_cq_s/read.